winlogbeat/sys/wineventlog/bufferpool.go

```diff
@@ -0,0 +1,113 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package wineventlog
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/v7/winlogbeat/sys"
+)
+
+// bufferPool contains a pool of byteBuffer objects.
+var bufferPool = sync.Pool{
+	New: func() interface{} { return &byteBuffer{buf: make([]byte, 1024)} },
+}
+
+// byteBuffer is an expandable buffer backed by a byte slice.
+type byteBuffer struct {
+	buf    []byte
+	offset int
+}
+
+// newByteBuffer return a byteBuffer from the pool. The returned value must
+// be released with free().
+func newByteBuffer() *byteBuffer {
+	b := bufferPool.Get().(*byteBuffer)
+	b.Reset()
+	return b
+}
+
+// free returns the byteBuffer to the pool.
+func (b *byteBuffer) free() {
+	if b == nil {
+		return
+	}
+	bufferPool.Put(b)
+}
+
+// Write appends the contents of p to the buffer, growing the buffer as needed.
+// The return value is the length of p; err is always nil. This implements
+// io.Writer.
+func (b *byteBuffer) Write(p []byte) (int, error) {
+	if len(b.buf) < b.offset+len(p) {
+		// Create a buffer larger than needed so we don't spend lots of time
+		// allocating and copying.
+		spaceNeeded := len(b.buf) - b.offset + len(p)
+		largerBuf := make([]byte, 2*len(b.buf)+spaceNeeded)
+		copy(largerBuf, b.buf[:b.offset])
+		b.buf = largerBuf
+	}
+	n := copy(b.buf[b.offset:], p)
+	b.offset += n
+	return n, nil
+}
+
+// Reset resets the buffer to be empty. It retains the same underlying storage
+// capacity.
+func (b *byteBuffer) Reset() {
+	b.offset = 0
+	b.buf = b.buf[:cap(b.buf)]
+}
+
+// Bytes returns a slice of length b.Len() holding the bytes that have been
+// written to the buffer.
+func (b *byteBuffer) Bytes() []byte {
+	return b.buf[:b.offset]
+}
+
+// Len returns the number of bytes that have been written to the buffer.
+func (b *byteBuffer) Len() int {
+	return b.offset
+}
+
+// SetLength resets the buffer to a specific length. It may allocate a new
+// underlying buffer discarding any existing contents.
+func (b *byteBuffer) SetLength(n int) {
```

Comment:
I understand this method is the main reason why you can't use a bytes.Buffer, am I right? I just find the name a bit misleading, what about Resize(n) or Reserve(n)?

Reply:
Both of those names are better. I'll go with `Reserve`.

The odd API comes out of its it dual use as both a generic byte array in Windows API calls and in its use as an `io.Writer` in the UTF16 to UTF8 conversions. I struggled with whether or not to leave this as it is. I'm sure another layer could be added to clean up the API.

I couldn't use bytes.Buffer because I needed direct access to the backing slice. And also in the past I found that simple wrapper around a slice was a little faster than bytes.Buffer. The result seems much closer now than I recall.

```
pkg: github.com/elastic/beats/winlogbeat/sys
BenchmarkByteBuffer/byteBuffer-12         	233508294	         4.98 ns/op
BenchmarkByteBuffer/bytes.Buffer-12       	154466978	         7.81 ns/op
PASS
```